mindmeld/auto_annotator.py

```diff
@@ -0,0 +1,691 @@
+        """ Converts an item returned from parse into a query entity.
+
+        Args:
+            item (dict): Dictionary representing an entity with the keys -
+                "body", "start", "end", "value", "dim".
+            processed_query (ProcessedQuery): The processed query that the
+                entity is found in.
+
+        Returns:
+            query_entity (QueryEntity): The converted query entity.
+        """
+        span = Span(start=item["start"], end=item["end"] - 1)
+        entity = Entity(text=item["body"], entity_type=item["dim"], value=item["value"])
+        query_entity = QueryEntity.from_query(
+            query=processed_query.query, span=span, entity=entity
+        )
+        return query_entity
+
+    def _resolve_conflicts(self, current_entities, annotated_entities):
+        """ Resolve overlaps between existing entities and newly annotad entities.
+
+        Args:
+            current_entities (list): List of existing query entities.
+            annotated_entities (list): List of new query entities.
+
+        Returns:
+            final_entities (list): List of resolved query entities.
+        """
+        overwrite = self.config["overwrite"]
+        base_entities = annotated_entities if overwrite else current_entities
+        other_entities = current_entities if overwrite else annotated_entities
+
+        additional_entities = []
+        for o_entity in other_entities:
+            no_overlaps = [
+                _no_overlap(o_entity, b_entity) for b_entity in base_entities
+            ]
+            if all(no_overlaps):
+                additional_entities.append(o_entity)
+        base_entities.extend(additional_entities)
+        return base_entities
+
+    # pylint: disable=R0201
+    def _unannotate_query(self, processed_query, remove_entities):
+        """ Removes specified entities in a processed query. If all entities are being
+        removed, this function will not remove entities that the annotator does not support
+        unless it is explicitly specified to do so in the config with the param
+        "unannotate_supported_entities_only" (boolean).
+
+        Args:
+            processed_query (ProcessedQuery): A processed query.
+            remove_entities (list): List of entities to remove.
+        """
+        remove_supported_only = self.config["unannotate_supported_entities_only"]
+        keep_entities = []
+        for query_entity in processed_query.entities:
+            if remove_entities == ["*"]:
+                is_supported_entity = self.valid_entity_check(query_entity.entity.type)
+                if remove_supported_only and not is_supported_entity:
+                    keep_entities.append(query_entity)
+            elif query_entity.entity.type not in remove_entities:
+                keep_entities.append(query_entity)
+        processed_query.entities = tuple(keep_entities)
+
+    @abstractmethod
+    def parse(self, sentence, **kwargs):
```

Comment:
we should implement the third abstract method

```
@property
@abstractmethod
def supported_entity_types(self):
   ...
```

this method will return the supported entity types for this particular annotator

Reply:
Sounds good. With this abstract method valid_entity_check() will not need to be implemented because valid_entity_check() will simply check if the given entity is in the list provided from the implementation of this abstract method.

Reply:
Added